src/miniscript/types/extra_props.rs

```diff
@@ -8,6 +8,90 @@ use MiniscriptKey;
 use Terminal;
 
 pub const MAX_OPS_PER_SCRIPT: usize = 201;
+// https://github.com/bitcoin/bitcoin/blob/9ccaee1d5e2e4b79b0a7c29aadb41b97e4741332/src/script/script.h#L39
+pub const HEIGHT_TIME_THRESHOLD: u32 = 500_000_000;
+
+/// Helper struct Whether any satisfaction of this fragment contains any timelocks
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Debug, Hash)]
+pub struct TimeLockInfo {
+    // csv with heights
+    pub csv_with_height: bool,
+    /// csv with times
+    pub csv_with_time: bool,
+    /// cltv with heights
+    pub cltv_with_height: bool,
+    /// cltv with times
+    pub cltv_with_time: bool,
+    /// combination of any heightlocks and timelocks
+    pub contains_combination: bool,
+}
+
+impl Default for TimeLockInfo {
+    fn default() -> Self {
+        Self {
+            csv_with_height: false,
+            csv_with_time: false,
+            cltv_with_height: false,
+            cltv_with_time: false,
+            contains_combination: false,
+        }
+    }
+}
+
+impl TimeLockInfo {
+    /// Whether the current contains any possible unspendable
+    /// path
+    pub fn contains_unspendable_path(self) -> bool {
+        self.contains_combination
+    }
+
+    // handy function for combining `and` timelocks
+    // This can be operator overloaded in future
+    pub(crate) fn comb_and_timelocks(a: Self, b: Self) -> Self {
+        Self::combine_thresh_timelocks(2, vec![a, b].into_iter())
+    }
+
+    // handy function for combining `or` timelocks
+    // This can be operator overloaded in future
+    pub(crate) fn comb_or_timelocks(a: Self, b: Self) -> Self {
+        Self::combine_thresh_timelocks(1, vec![a, b].into_iter())
+    }
+
+    pub(crate) fn combine_thresh_timelocks<I>(k: usize, sub_timelocks: I) -> TimeLockInfo
+    where
+        I: IntoIterator<Item = TimeLockInfo>,
+    {
+        let mut timelock_info = TimeLockInfo::default();
+        // timelocks calculation
+        // k >= 2 implies it is possible to have timelocks and heightlocks
+        // merged together
+        // timelocks.csv_with_height and timelocks.csv_with_time can be directly
+        // combined together instead of iterating through all the pairs:
+        // consider the two cases:
+        // 1) Only one subfragment contains timelock and hieghtlock: In this case,
+        // timelocks.contains_combination would already by `true`. and other combinations would
+        // not add any effect.
+        // 2) hieghtlocks and timelocks are distributed across two branches:
+        // and we check if any of the existing branches has correponding lock
+        // that cannot be combined
+        // If k == 1, then this is no combination possible.
```

Comment:
I find this whole comment block confusing. I think you should replace it with
```
// Propagate all fields of `TimelockInfo` from each of the node's children to the node
// itself (by taking the logical-or of all of them). In case `k == 1` (this is a disjunction)
// this is all we need to do: the node may behave like any of its children, for purposes
// of timelock accounting.
//
// If `k > 1` we have the additional consideration that if any two children have conflicting
// timelock requirements, this represents an inaccessible spending branch. If there are
// two such children we must set the `contains_combination` field.
```